### PR TITLE
Don't allocate in `associated_type_shorthand_candidates`

### PR DESCRIPTION
**Pull request by Veykril**

### CODE REVIEW

File: crates/hir_ty/src/lower.rs [hunk 1]
```diff
@@ -821,24 +821,37 @@ pub fn associated_type_shorthand_candidates<R>(
...
-                    _ => None,
-                })
-                .collect();
+            if let res @ Some(_) = predicates.iter().find_map(|pred| match &pred.value.value {
+                // FIXME: how to correctly handle higher-ranked bounds here?
+                WhereClause::Implemented(tr) => {
+                    search(tr.clone().shift_bound_vars_out(DebruijnIndex::ONE))
+                }
+                _ => None,
+            }) {
```
Comment (matklad) [inline crates/hir_ty/src/lower.rs:852]: Feels like it can benefit from pulling into the variable.

Comment (matklad) [reply crates/hir_ty/src/lower.rs:852]: https://github.com/rust-analyzer/rust-analyzer/pull/8293

**Comment by Veykril** [comment]
changelog skip

**Pull request merged**